### DISCUSSION TITLE
TypeError: projects.map is not a function

### DISCUSSION THREAD

**Discussion by BenJarrett** [nss-evening-cohort-14/evening-client-side#252] [Help Me!] [answered]
Ben Jarrett

#162

When trying to print my project cards in my project route, I keep getting this error. 
I have seen the ".map is not a function" error before, I just cannot figure out what is causing it this time. 

TypeError: projects.map is not a function

https://stackoverflow.com/questions/44574367/react-map-is-not-a-function
https://forum.freecodecamp.org/t/map-is-suddenly-not-a-function/134196
https://www.pluralsight.com/guides/typeerror-handling-in-react.js-for-map-function

I have tried a few different things. 
I've tried wrapping my used imports and exports in brackets, tried re-writing my project card and projectForm components, drilling some props, and removing others. I can not quite crack this and it has put me way before where I want to be right now. I feel like once I get past this I can get crud done pretty confidently. 

https://github.com/BenJarrett/personal-bio-site-react/tree/projectRead

**Comment by drteresavasquez** [comment]
It probably means that projects is not an array. Check the data type and make sure you're mapping an array.

**Reply by BenJarrett** [reply]
Okay, I'll check that now. 

Thank you!

**Reply by BenJarrett** [reply]
``` import React, { useState, useEffect } from 'react';
import firebase from 'firebase';
import Routes from '../helpers/Routes';
import './App.scss';
import NavBar from './components/NavBar';
import { getProjects } from '../helpers/data/projectsData';

function App() {
  const [admin, setAdmin] = useState(null);
  const [projects, setProjects] = useState([]);

  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
        getProjects().then((projectsArray) => setProjects(projectsArray));
      } else if (admin || admin === null) {
        setAdmin(false);
        getProjects().then((projectsArray) => setProjects(projectsArray));
      }
    });
  }, []);

  return (
    <>
      <NavBar admin={admin} />
      <Routes
       projects={projects}
       setProjects={setProjects}
      />
      </>
  );
}

export default App;

**Reply by BenJarrett** [reply]
This is what my app/index.js looks like where I am setting my projects to an empty array.
I feel like I'm going crazy trying to figure this out. Been going at it since last night. 

Is there something I am missing while I'm banging my head against a wall on this error?

**Reply by BenJarrett** [reply]
I removed the error, changed it to a spread operator. I am printing one card at the moment, but with no values. 
I have confirmed that on my axios call, I am getting an array of objects. 
I am setting the projects/setProjects useState to an empty array. 
I have confirmed that my projects propType is an array. 

Somewhere from my data to my .map/spread operator, it's becoming undefined. 
Numerous class time and extra hours been spent trying to debug this and I cannot figure out what is going on.

**Reply by drteresavasquez** [reply]
Make sure you're passing it where you need it. Undefined means it doesn't exist.

**Reply by BenJarrett** [reply]
I could have sworn that I checked 15 times that Projects is passing everywhere it needs to be, but I’ve been trying to figure this out for days and my eyes might not have caught it! 

When I get home I will check all of that again. I just wanna get this error figured out so I can move on with the site.

thank you!

**Reply by BenJarrett** [reply]
Just an update:

Still not seeing where I am not passing projects as an array. 
I’ll keep trying to figure this out

**Reply by drteresavasquez** [reply]
Also, you can't just refresh the app on the view you're on if projects are being set in app/index.js

Try starting your app from the beginning and navigate to where you have projects. If this works I would recommend moving state to where it is needed rather than passing it down.

**Comment by BenJarrett** [comment, marked as answer]
OMG Y'ALL...

A curly bracket.. A curly bracket is what I was missing. 
I have been troubleshooting this for DAYS. 
Thank you to my main mans Rob for looking at it with fresh eyes and catching this. 

For the love of all that is good in life... 
PUT CURLY BRACKETS AROUND YOUR PROPS
```
export default function Routes({
  admin,
  projects,
  setProjects,
  firebaseKey,
  githubUrl,
  screenshot,
  technologiesUsed,
  title,
  url,
})
